File: mac_app/Utilities/TranscriptDeduplicator.swift
import Foundation

/// Filters microphone "bleed" — when the mic picks up remote participants' audio
/// from the speakers, causing the same speech to be transcribed on both streams.
///
/// Strategy: Keep a rolling window of recent system audio transcripts. When a mic
/// transcript arrives, compare it against the window. If word overlap exceeds the
/// threshold, it's bleed (the mic heard the speakers) → drop it.
@MainActor
final class TranscriptDeduplicator {
    // MARK: - Configuration

    /// How long to keep system transcripts for comparison (seconds)
    private let windowDuration: TimeInterval = 5.0

    /// Minimum word overlap ratio to consider a mic transcript as bleed
    private let similarityThreshold: Double = 0.4

    /// Minimum word count to attempt dedup (very short phrases are too ambiguous)
    private let minWordsForDedup = 3

    // MARK: - State

    private struct RecentEntry {
        let words: Set<String>
        let timestamp: Date
    }

    private var recentSystemTranscripts: [RecentEntry] = []

    // MARK: - Public API

    /// Record a system audio transcript for future bleed comparison.
    /// Call this EVERY time a system audio transcript is received.
    func addSystemTranscript(_ text: String) {
        let words = Self.normalizeWords(text)
        guard !words.isEmpty else { return }

        recentSystemTranscripts.append(RecentEntry(words: words, timestamp: Date()))
        pruneOldEntries()
    }

    /// Check if a mic transcript is likely bleed from the speakers.
    /// Returns `true` if the text matches a recent system audio transcript → should be dropped.
    func isMicTranscriptBleed(_ text: String) -> Bool {
        pruneOldEntries()

        let micWords = Self.normalizeWords(text)

        // Don't dedup very short phrases — too ambiguous
        guard micWords.count >= minWordsForDedup else { return false }

        for entry in recentSystemTranscripts {
            let commonWords = micWords.intersection(entry.words)
            let maxCount = max(micWords.count, entry.words.count)
            let similarity = Double(commonWords.count) / Double(maxCount)

            if similarity >= similarityThreshold {
                print("[Dedup] Bleed detected (\(Int(similarity * 100))% overlap): mic=\(micWords.count) words, sys=\(entry.words.count) words")
                return true
            }
        }

        return false
    }

    /// Reset all state (call on session end)
    func reset() {
        recentSystemTranscripts.removeAll()
    }

    // MARK: - Private

    private func pruneOldEntries() {
        let cutoff = Date().addingTimeInterval(-windowDuration)
        recentSystemTranscripts.removeAll { $0.timestamp < cutoff }
    }

    /// Normalize text into a set of lowercase words, stripping punctuation
    private static func normalizeWords(_ text: String) -> Set<String> {
        let cleaned = text.lowercased()
            .components(separatedBy: CharacterSet.alphanumerics.inverted)
            .filter { $0.count > 1 } // Drop single-char noise
        return Set(cleaned)
    }
}

File: win_app_2/src/services/transcription/transcriptDeduplicator.ts
/**
 * TranscriptDeduplicator - Filters microphone "bleed"
 *
 * When the mic picks up remote participants' audio from the speakers,
 * the same speech gets transcribed on both streams. This service detects
 * and filters the duplicate mic transcripts.
 *
 * Three-layer strategy (aligned with macOS TranscriptDeduplicator.swift):
 * 1. Temporal suppression: if system audio produced a transcript within 1.5s, suppress ALL mic transcripts
 * 2. Word overlap: compare mic words against recent system transcripts (35% threshold, 10s window)
 * 3. Session-wide: if 70%+ of mic words already appeared in system audio during this session
 */

import { createLogger } from '@/lib/logger'

const log = createLogger('Dedup')

// Configuration (aligned with macOS values)
const WINDOW_DURATION = 10_000 // 10s — how long to keep system transcripts for comparison
const TEMPORAL_SUPPRESSION_WINDOW = 1500 // 1.5s — suppress mic while system audio is active
const SIMILARITY_THRESHOLD = 0.35 // 35% word overlap = bleed
const MIN_WORDS_FOR_DEDUP = 2 // minimum word count to attempt dedup

interface RecentEntry {
  words: Set<string>
  timestamp: number
}

let recentSystemTranscripts: RecentEntry[] = []
let lastSystemTranscriptTime: number | null = null
let allSessionSystemWords: Set<string> = new Set()

/** Normalize text into a set of lowercase words, stripping punctuation */
function normalizeWords(text: string): Set<string> {
  const words = text
    .toLowerCase()
    .split(/[^a-zA-ZÀ-ÿ0-9]+/)
    .filter((w) => w.length > 1) // Drop single-char noise
  return new Set(words)
}

function pruneOldEntries(): void {
  const cutoff = Date.now() - WINDOW_DURATION
  recentSystemTranscripts = recentSystemTranscripts.filter((e) => e.timestamp > cutoff)
}

/**
 * Record a system audio transcript for future bleed comparison.
 * Call this EVERY time a system audio transcript (interim or final) is received.
 */
export function addSystemTranscript(text: string): void {
  const words = normalizeWords(text)
  if (words.size === 0) return

  lastSystemTranscriptTime = Date.now()
  recentSystemTranscripts.push({ words, timestamp: Date.now() })
  words.forEach((w) => allSessionSystemWords.add(w))
  pruneOldEntries()
}

/**
 * Check if a mic transcript is likely bleed from the speakers.
 * Returns true if the text should be dropped.
 */
export function isMicTranscriptBleed(text: string): boolean {
  pruneOldEntries()

  // Layer 1: Temporal suppression
  if (lastSystemTranscriptTime !== null) {
    const elapsed = Date.now() - lastSystemTranscriptTime
    if (elapsed < TEMPORAL_SUPPRESSION_WINDOW) {
      log.info(`Temporal suppression (system audio active ${(elapsed / 1000).toFixed(1)}s ago): "${text.substring(0, 60)}"`)
      return true
    }
  }

  const micWords = normalizeWords(text)
  if (micWords.size < MIN_WORDS_FOR_DEDUP) return false

  // Layer 2: Word overlap against recent window
  for (const entry of recentSystemTranscripts) {
    const common = new Set([...micWords].filter((w) => entry.words.has(w)))
    const maxCount = Math.max(micWords.size, entry.words.size)
    const similarity = common.size / maxCount

    if (similarity >= SIMILARITY_THRESHOLD) {
      log.info(`Word overlap (${Math.round(similarity * 100)}%): mic=${micWords.size} words, sys=${entry.words.size} words`)
      return true
    }
  }

  // Layer 3: Session-wide word overlap
  if (allSessionSystemWords.size > 0) {
    const common = new Set([...micWords].filter((w) => allSessionSystemWords.has(w)))
    const ratio = common.size / micWords.size
    if (ratio >= 0.7) {
      log.info(`Session-wide overlap (${Math.round(ratio * 100)}%): ${common.size}/${micWords.size} mic words seen in system audio`)
      return true
    }
  }

  return false
}

/** Reset all state (call on session end) */
export function reset(): void {
  recentSystemTranscripts = []
  lastSystemTranscriptTime = null
  allSessionSystemWords = new Set()
}
